[wip] slack_import: Improve how Slack thread messages are handled.

PieterCK opened the pull request:
Stacked on top of #35382

---

## Overview
This PR enhances the handling of Slack thread messages during the Slack export conversion for import. Additionally, it introduces logic to address near-link messages in the import process. Note that some specifications differ from the original details mentioned in the linked GitHub issues due to discussions in the [CZO discussion](https://chat.zulip.org/#narrow/stream/138-user-questions/topic/What.20to.20do.20with.20Slack.20imported.20topics.3F) and subsequent iterations.

Fixes #27626.

Original thread messages are now placed in the main import topic, with an additional cross-linking message referencing its thread topic.
Added support to convert thread replies marked with "Also send to #channel". 

thread topics and their original thread messages:
![image](https://github.com/user-attachments/assets/c4cdbbf1-25d8-40a8-8ba6-c01bbd263d11)

Converted Slack thread. 
| Slack | Zulip |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/3eef77e8-edf9-4bdb-9883-482440d60c65) | ![image](https://github.com/user-attachments/assets/f37b1be5-e2ef-4049-94a3-f5d8ba0d9faf) |

the original thread message is now in the main "imported from Slack" topic instead of in its branched-off thread topic. 
| Slack | Zulip |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/57c55491-a07e-4fb0-ac8d-de01aa28da24) | ![image](https://github.com/user-attachments/assets/ab3b8ad6-8050-4d80-b167-3170f64dd673) |

Thread reply marked with "Also send to #channel".
| Slack |
| --- |
| ![image](https://github.com/user-attachments/assets/47092b16-0e67-4d8b-bbc6-e06c9a129ee9)  |

| Zulip |
| --- |
| ![image](https://github.com/user-attachments/assets/879d5457-7f20-4659-a3b1-65f9ba2e9375) |

Thread conversations with deleted or missing original thread messages.
![image](https://github.com/user-attachments/assets/e34ced6f-a5d6-477a-9e62-f536acc72df0)

Imported thread topics.
![image](https://github.com/user-attachments/assets/7c70fa79-7bf2-4a55-9be2-f5f6a654109e)




[CZO](https://chat.zulip.org/#narrow/stream/138-user-questions/topic/What.20to.20do.20with.20Slack.20imported.20topics.3F)


## Follow Up:
</summary>

- Bumping an old Rocket Chat thread/reply conversion feature here. A couple of users have requested an implementation of a similar thread/reply conversion for Rocket Chat export. [CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Rocketchat.20import/near/1210484)
- Once this PR is merged, changes related to #30465 can be refactored to use the new `get_zulip_mention_for_slack_user` to reduce duplicative code.
</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
slack_importer: Add cross-linking message for Slack thread.

This commit sends the original thread message to the main import topic
instead of dumping it in together with its replies in a separate thread
topic.

It also appends a cross-linking notification message to the original
thread message, linking it to its thread topic.

Fixes part of #27626.

PieterCK committed:
slack_importer: Add thread message broadcast feature.

This commit adds a new logic to handle thread messages
that are marked with "Also send to #channel". This Slack
feature is mirrored by echoing the marked thread reply
message to the main import topic.

Fixes part of #27626.

PieterCK committed:
slack_importer: Add quote-and-reply to first thread reply.

This makes the converted threads' first reply have quote-and-reply to
the original thread message.

PieterCK commented:
@zulipbot add "buddy review"

@kennethnrk Please review this! Thanks

PieterCK commented:
@zulipbot remove "buddy review"

@kennethnrk - Hi, I'll untag the buddy review label for this PR for now because I'll be adding fix for a related [issue](https://chat.zulip.org/#narrow/stream/3-backend/topic/exporting.20.26.20importing.20messages.20with.20.2Fnear.2F.20link/near/1902081) here

PieterCK commented:
@kennethnrk - Could you take a look at this PR? Thanks! Let me know if you need help with anything

PieterCK commented:
@zulipbot add "mentor review"

@sbansal1999 I think we can do mentor review in parallel for this one, what do you think?

kennethnrk commented:
All three fixes work as expected, LGTM @PieterCK 

@zulipbot remove "buddy review"

PieterCK commented:
@kennethnrk  - Thanks for checking it out

alya commented:
@sbansal1999 any feedback on this one?

alya commented:
@timabbott will take a look at this next.

PieterCK commented:
I split out the fix for broken near-links in imported messages into this PR:
- #32272

PieterCK commented:
I rebased this to be independent of #32272. This is up for review!

PieterCK commented:
I've updated this, and it's ready for review again. I did manual testing again, but didn't update the screenshots in the PR because they're not outdated and show more details and cases. Screenshot from quick test:
| Zulip | Slack |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/5d63c327-3672-4858-b52a-7671504b2c12) | ![image](https://github.com/user-attachments/assets/f97882ab-aeb4-459b-a90f-928c22d7e100) ![image](https://github.com/user-attachments/assets/37c44d70-28c4-40b5-8f74-e871a7349a12)|

PieterCK commented:
I've split a commit that'll fix https://github.com/zulip/zulip/issues/27661 and a prep commit for this PR to #35151.

PieterCK committed:
test_slack_importer: Re-focus Slack thread conversion tests.

This re-organizes several tests to reduce busy work when making
adjascent changes.
